sqlglot/optimizer/qualify_columns.py

```diff
@@ -0,0 +1,321 @@
+        parent_selectables,
+        _qualify_derived_tables(ctes, schema, sequence, parent_selectables, chain=True),
+    )
+    selectables = _merge({}, _qualify_tables(tables, schema, parent_selectables))
+    selectables = _merge(
+        selectables,
+        _qualify_derived_tables(derived_tables, schema, sequence, parent_selectables),
+    )
+    expansions = _expand_stars(select_stars, selectables)
+    columns.extend(expansions)
+    selections.extend(expansions)
+    selectables = _merge(parent_selectables, selectables)
+    _qualify_columns(columns, selectables)
+    _qualify_subqueries(subqueries, schema, sequence, selectables)
+    return _qualify_outputs(selections)
+
+
+def _qualify_derived_tables(
+    derived_tables, schema, sequence, parent_selectables, chain=False
+):
+    selectables = {}
+    for subquery in derived_tables:
+        subquery_outputs = _qualify_union(
+            expression=subquery.this,
+            schema=schema,
+            sequence=sequence,
+            parent_selectables=_merge(parent_selectables, selectables)
+            if chain
+            else parent_selectables,
+        )
+
+        table_alias = subquery.args.get("alias")
+        if not table_alias:
+            table_alias = exp.TableAlias()
+            subquery.set("alias", table_alias)
+
+        alias = table_alias.args.get("this")
+        if not alias:
+            alias = exp.to_identifier(f"_alias{sequence.next()}")
+            table_alias.set("this", alias)
+
+        alias_columns = table_alias.args.get("columns")
+        if not alias_columns:
+            alias_columns = [exp.to_identifier(o) for o in subquery_outputs]
+            table_alias.set("columns", alias_columns)
+
+        selectable_table = alias.text("this")
+
+        selectable_columns = [c.text("this") for c in alias_columns]
+        selectables = _merge(selectables, {selectable_table: selectable_columns})
+
+    return selectables
+
+
+def _qualify_tables(tables, schema, parent_selectables):
+    selectables = {}
+    for table in tables:
+        table_name = table.text("this")
+
+        if isinstance(table.parent, exp.Alias):
+            selectable_table = table.parent.alias
+        else:
+            selectable_table = table_name
+
+        if table_name in parent_selectables:
+            selectable_columns = parent_selectables[table_name]
+        else:
+            selectable_columns = list(schema[table_name])
+
+        selectables = _merge(selectables, {selectable_table: selectable_columns})
+    return selectables
+
+
+def _expand_stars(select_stars, selectables):
+    result = []
+    for star in select_stars:
+        new_columns = []
+
+        if isinstance(star, exp.Column):
+            tables = [star.text("table")]
+        else:
+            tables = list(selectables)
+
+        for table in tables:
+            columns = selectables.get(table, [])
+            for column in sorted(columns):
+                new_columns.append(
```

Comment:
is this cleaner than

parse_one(f'"{table}"."{column}"')

Reply:
They feel similar to me.
I guess I don't always trust `parse` to give me what I want... so the more explicit instantiation feels nice. But I don't have a strong opinion here.

Reply:
Er, parse_one(f'"{table}"."{column}"') turns:

```sql
SELECT x.a AS a, x.b AS b FROM x
```

into 

```sql
SELECT "x"."a" AS a, "x"."b" AS b FROM x
```

If we go that route, maybe we should be quoting _everything_?

Reply:
canonical form should quote everything, but i sort of have this with sql(identify=true)